### PR TITLE
Change: Big UFO disaster targets current location of a random train

### PR DESCRIPTION
**Pull request by LordAro**
## Motivation / Problem
See #10076
Closes #10076

## Description
As suggested, Big UFO now targets the current location of a random train.
Code largely copied from the standard UFO - differs in that it uses the current tile of the train, rather than the index, because we're not actively targeting a train

## Limitations
Limited testing, because disasters are a pain to spawn. Potential feature: Cheat to spawn a disaster.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/disaster_vehicle.cpp [hunk 1]
```diff
@@ -578,17 +579,27 @@ static bool DisasterTick_Big_Ufo(DisasterVehicle *v)
...
+
+		if (n == 0) {
+			/* If there are no targetable trains, destroy the UFO. */
+			delete v;
+			return false;
+		}
+
+		n = RandomRange(n); // Choose one of them.
+		for (const Train *t : Train::Iterate()) {
+			/* Find (n+1)-th road vehicle. */
```
Comment (stormcone) [inline src/disaster_vehicle.cpp]:
```suggestion
			/* Find (n+1)-th train. */
```

Comment (LordAro) [reply src/disaster_vehicle.cpp]: Can you tell the code was copied?

Comment (stormcone) [reply src/disaster_vehicle.cpp]: I actually read it in the description. :)

File: src/disaster_vehicle.cpp [hunk 2]
```diff
@@ -578,17 +579,27 @@ static bool DisasterTick_Big_Ufo(DisasterVehicle *v)
 		}
 		v->current_order.SetDestination(1);
 
-		TileIndex tile_org = RandomTile();
-		TileIndex tile = tile_org;
-		do {
-			if (IsPlainRailTile(tile) &&
-					Company::IsHumanID(GetTileOwner(tile))) {
```
Comment (glx22) [inline src/disaster_vehicle.cpp:585]: AIs are not really able to manage Big UFO effects, that's why only human owned tiles were targeted.
I think the new method should also filter out AI owned vehicles.

**Comment by JGRennison** [comment]
On the testing side, you could probably locally cherry pick this: https://github.com/JGRennison/OpenTTD-patches/commit/4cdc253c38fe7378e2e9cf0ff58bf2e4bc0baa68

**Review by rubidium42 — DISMISSED** [review]

**Comment by nielsmh** [comment]
Shouldn't this also test if the train is actually running and not in a depot?
I suppose it's not a real problem, and targeting depots is valid too. (Do trains in depot have their correct tile stored?)

**Comment by glx22** [comment]
Yeah skipping crashed and hidden vehicles (in depot, tunnel) might be a good idea.

**Comment by rubidium42** [comment]
> Yeah skipping crashed and hidden vehicles (in depot, tunnel) might be a good idea.

Though then you also have to do the same for road vehicles, so they are consistent. And then the scope of the change grows significantly.

**Comment by glx22** [comment]
For road vehicle the target is the vehicle itself, it can move while the UFO is coming. Here we just want a tile, and the IsPlainRailTile() check has been removed.

**Comment by JGRennison** [comment]
> > Yeah skipping crashed and hidden vehicles (in depot, tunnel) might be a good idea.
> 
> Though then you also have to do the same for road vehicles, so they are consistent. And then the scope of the change grows significantly.

The various disasters are extremely arbitrary. There isn't some underlying consistency to be exposed and any possible changes to small UFO road vehicles don't have to be done in the same PR.

**Comment by LordAro** [comment]
¯\\\_(ツ)_/¯

**Review by glx22 — APPROVED** [review]

**Pull request merged**